inlang/source-code/versioned-interfaces/project-settings/src/interface.ts

```diff
@@ -76,9 +76,7 @@ const InternalProjectSettings = Type.Object({
 			],
 		})
 	),
-	featureFlags: Type.Optional(
-		Type.Record(Type.String(), Type.Union([Type.Boolean(), Type.Number(), Type.String()]))
-	),
+	featureFlags: Type.Optional(Type.Record(Type.String(), Type.Boolean())),
```

Comment:
@samuelstroschein i was going for ease of manual testing (just toggle the flag between true and false without having to lookup the flag name and recreate the line) but I don't mind forcing it to literal "true" if that encourages folks to remove old flags.
```ts
	featureFlags: Type.Optional(Type.Record(Type.String(), Type.Literal(true))),
```

Here's a CLI experience after changing the value from `true` to `false` 

```
$ inlang machine translate --project project.inlang/

 WARN  Human translations are better than machine translations.                                                                                                  5:08:39 PM

We advise to use machine translations in the build step without commiting them to the repo. By using machine translate in the build step, you avoid missing translations in production while still flagging to human translators that transaltions are missing. You can use the force flag (-f, --force) to skip this prompt warning.

✔ Are you sure you want to machine translate? … yes

 ERROR  The project settings are invalid:                                                                                                                        5:08:41 PM

"/featureFlags/aliases":

Expected true

  
  "/featureFlags/aliases":
  
  Expected true
  at parseSettings (/Users/jldec/inlang/monorepo/inlang/source-code/cli/dist/main.js:109522:13)
  at loadSettings (/Users/jldec/inlang/monorepo/inlang/source-code/cli/dist/main.js:109515:10)
```

Reply:
the testing one is a good argument at the cost of outdated feature flags. what trade off should we do here? 

- we don't want people to manually edit .inlang files (argument against testing stuff) 
- an automatic clean up of outdated feature flags will be just like just as annoying for testing than removing the key entirely
- have a manual clean up functions requires apps to build show feature

an alternative might be a map/array to avoid the weird setting to true thing

```
featureFlags: [
  "aliases",
  "humanIds"
]
```

```ts
if (project.settings.featureFlags.includes("aliases"){

}
```

@jldec what do you think about the array/map approach?

Reply:
I prefer the map, even with the forced true (maps are more easily tested with `featureflags?.foo`)
I don't think we really gain a lot by preventing "false" - we'd gain more by maintaining the list of valid feature flags somewhere and validating against that - but I don't want to do it in the interface declaration because that seems too hard to release and rollout changes (maybe I'm missing the process for that).

Reply:
> we'd gain more by maintaining the list of valid feature flags somewhere and validating against that

We will likely not be able to validate the feature flags:

- apps will use the feature to drop their own feature flags into the settings 
- an old SDK version can't crash once new feature flags are added to the validation list



@jldec i vote for only setting to `true` for now. we can loosen the restriction in the future. the beauty of allowing only `true` is a) automatic removal of unused feature flags and b) one way to achieve things:  

**Comparing both**

1. loading a project can get slower: 100 feature flags increase the size of the settings json by approx 1.5kb. assuming that apps will also use the feature flag feature, i easily foresee 100's of feature flags "rotten" in the settings file of which only a few are active.

2. multiple ways to achieve the same thing (remove a key or set it to false. which method should be preferred?)  

```ts

// 2 ways to achieve the same things :/ 
project.settings.featureFlags.aliases = false
project.settings.featureFlags.aliases = undefined

```

**Conclusion**

can we agree on: 

1. `settings.featureFlags` should not be undefinable to increase ergonomics of using the API (have a migration that creates sets feature flags to `{}` if not existent).
2. have a map instead of an array
3. only allow `true` for now. can be loosened in the future. 
4. have a good error message when a feature flag is set to false that prompts the user to remove the key entirely.

Reply:
not sure how to implement 4, but will work on 1-3

Reply:
> have a good error message when a feature flag is set to false 

The more important challenge will to clean up code and config for feature flags that are no longer needed. Migration of settings to auto-remove those would be a big win (future.)

I can take another looking, but I'm not convinced that we should add extra code now to warn users (or change the default error message) when the value of the flag is set to false.

Reply:
> When we ship the new project-settings schema, old SDKs and old projects, need to keep loading without error - we don't want to force everyone to upgrade (break their projects) with this PR.

Argh. I forgot about this. 

In the future, we should have a public telemetry dashboard (cc @jannesblobel) on API usage of the SDK over time and formulate a rule like "if a deprecated API is used by less than X% of projects, it will be removed with the next major version". I hope that such a public dashboard aligns our users with deprecations. 

In this example, we could formulate an insight/rule/deprecation like "once less than 5% of projects use SDK version <X (which includes the featureFlag migration), feature flags can not be undefined"